Address Set-Location/Get-Item Issue With ?/* Characters

NoMoreFood opened the pull request:
# PR Summary

- Adjusted behavior of GetCorrectCasedPath() which attempts to search for the file on the file system in order to normalize its case.  Since some file systems support wildcard-type characters (e.g. MacOS) in path names so these need to be skipped to avoid erroneous resolution.
- Added tests to detect this behavior.

## PR Context

This corrects the behavior where Set-Location fails to change directory or changes to an entirely different child directory when the target directory contains characters such as '*'.  Similarly, Get-Item also returns errant information.

Resolves: https://github.com/PowerShell/PowerShell/issues/12299

## PR Checklist

- [X] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [X] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [X] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): <!-- Experimental feature name(s) here -->
- **User-facing changes**
    - [X] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: <!-- Number/link of that issue here -->
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [X] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [X] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

NoMoreFood committed:
Address Set-Location & Get-Item Issue With ?/* Characters

- Adjusted behavior of GetCorrectCasedPath() which attempts to search for the file on the file system in order to normalize its case. Since some file systems support wildcard characters in path names; processing needs to be skipped for these.
- Added tests to detect this behavior.

NoMoreFood commented:
@TravisEz13 So I can categorize these correctly in the future, can you educate me as to why this would be considered a breaking change?

NoMoreFood commented on test/powershell/Modules/Microsoft.PowerShell.Management/Set-Location.Tests.ps1:
```diff
@@ -231,3 +231,58 @@ Describe "Set-Location" -Tags "CI" {
+
+        $dirFullName = (Join-Path $TestDrive $Name) 
+        
+        $newDir = [System.IO.Directory]::CreateDirectory($dirFullName)
```
Actually, the reason I didn't do it this way is because `New-Item -Type Directory -Name 'example '` erroneously truncates the extra space on Linux/MacOS and it shoudn't.  I'll put in a separate issue to track that issue.

NoMoreFood commented on test/powershell/Modules/Microsoft.PowerShell.Management/Set-Location.Tests.ps1:
```diff
@@ -231,3 +231,58 @@ Describe "Set-Location" -Tags "CI" {
+        # is quite broken in many powershell cmdlets
+        if ($Name.Contains('\')) 
+        {
+            Set-ItResult -Skipped -Because "path elements with backslashes are not fully supported in PowerShell on this operating system."
```
It's only MacOS / Linux.  I'll clarify the comment and change to 'Pending'.

iSazonov commented on src/System.Management.Automation/namespaces/FileSystemProvider.cs:
```diff
@@ -158,6 +158,11 @@ private static string GetCorrectCasedPath(string path)
                         // This handles short path names
                         exactPath += StringLiterals.DefaultPathSeparator + item;
                     }
+                    else if (item.IndexOfAny(new char[] { '*', '?' }) >= 0)
```
We can re-use Utils.Separators.StarOrQuestion static value.

NoMoreFood replied:
Excellent; I was wondering if there was a static value somewhere.  Thanks!

SteveL-MSFT commented:
@PowerShell/powershell-committee agrees that the current behavior is broken where:

```powershell
set-location -literalpath ?
```

must treat `?` as a literal value and not a wildcard.  We don't believe this is a breaking change that causes concern, however, we are concerned with the location of the code change in a helper that is used by many code paths which itself may introduce a regression somewhere else.

NoMoreFood committed:
Address Set-Location & Get-Item Issue With ?/* Characters - Revisions

- Mark backslash test on MacOS / Linux as 'Pending' vice 'Skipped'.
- Use Utils.Separators.StarOrQuestion instead of hard-coded constants.

NoMoreFood commented on test/powershell/Modules/Microsoft.PowerShell.Management/Set-Location.Tests.ps1:
```diff
@@ -231,3 +231,58 @@ Describe "Set-Location" -Tags "CI" {
+
+        try {
+            Set-Location -LiteralPath $dirFullName
+            [System.IO.Path]::GetFileName((Get-Location).Path) | Should -BeExactly $Name
```
Already issue filed: https://github.com/PowerShell/PowerShell/issues/8751

NoMoreFood commented:
@TravisEz13 Sorry, the day job has distracted me from following through on this and one other PR.  I honestly thought the "location of the code change" comment was sort of meant to be rhetorical.  The bug itself can be scoped to that helper function: In what scenario would I possibly pass a literally, existing directory path to GetCorrectCasedPath() and it would be acceptable to get back a reference to an existing different directory?  The function exists solely to normalize case within a path --- it should NEVER be returning a path to a different file.

NoMoreFood commented on test/powershell/Modules/Microsoft.PowerShell.Management/Set-Location.Tests.ps1:
```diff
@@ -231,3 +231,58 @@ Describe "Set-Location" -Tags "CI" {
+        ForEach ($specialChar in $specialChars) {
+            If ($specialChar -ne '.') { $testCases += @{Name=([string] $specialChar)} }
+            $testCases += @{Name=('foo' + $specialChar)}
+            $testCases += @{Name=($specialChar + 'foo')}
```
So for #2, you're thinking a test where something where 'a???b' exists literally on the filesystem and we make sure that a lookup on 'a111b' doesn't match it?

NoMoreFood replied:
Tests added.

NoMoreFood committed:
Address Set-Location & Get-Item Issue With ?/* Characters - Revisions

- Added failure test for a directory that contains wildcard characters.

iSazonov commented on src/System.Management.Automation/namespaces/FileSystemProvider.cs:
```diff
@@ -158,6 +158,11 @@ private static string GetCorrectCasedPath(string path)
+                    else if (item.IndexOfAny(Utils.Separators.StarOrQuestion) >= 0)
+                    {
+                        // This handles literal wildcard characters that could resolve erroneously
+                        exactPath += StringLiterals.DefaultPathSeparator + item;
```
The line is the same as 159 - we could combine both condition blocks.

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Management/Set-Location.Tests.ps1:
```diff
@@ -230,4 +230,79 @@ Describe "Set-Location" -Tags "CI" {
         }
     }
+
+    It 'Should should not match directory containing wildcard character *' -Skip:([System.IO.Path]::GetInvalidFileNameChars() -contains '*') {
```
```suggestion
    It 'Should not match directory containing wildcard character *' -Skip:([System.IO.Path]::GetInvalidFileNameChars() -contains '*') {
```

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Management/Set-Location.Tests.ps1:
```diff
@@ -230,4 +230,79 @@ Describe "Set-Location" -Tags "CI" {
+            [System.IO.Path]::GetFileName((Get-Location).Path) | Should -BeExactly $Name
+        }
+        catch {
+            throw "Set-Location $Name Failed"
```
Why do we hide original exception?

NoMoreFood committed:
Address Set-Location & Get-Item Issue With ?/* Characters - Revisions

iSazonov commented on src/System.Management.Automation/namespaces/FileSystemProvider.cs:
```diff
@@ -153,14 +153,9 @@ private static string GetCorrectCasedPath(string path)
                         break;
                     }
-                    else if (item.Contains('~'))
+                    else if (item.Contains('~') || item.IndexOfAny(Utils.Separators.StarOrQuestion) >= 0)
```
We could create new static variable for all 3 chars.

NoMoreFood replied:
I'm not sure if this single instance merits it.  I don't foresee many instances where this subset of special characters would share a common categorization on a regular basis.  Your call.

NoMoreFood commented:
Is this ready for merge or are we still waiting for reviews?

NoMoreFood reviewed (commented):
I believe I have resolved these.

iSazonov commented:
The GetCorrectCasedPath() is not in right place. This would only have to be used in Formatting System and completors.

NoMoreFood commented:
@TravisEz13 All of the non-file system paths you reference end up calling a base class version or provider-specific version of NormalizePath() so this change luckily does not affect them.

@iSazonov I think that might be splitting hairs to some degree.  When somebody does a Set-Location to a specific path, do you want the resultant path at the prompt to be case-accurate or not?  I think it might be hard to address all the "what if" sort of cases.  Regardless, this PR was authored to address the "bug" within GetCorrectCasedPath() itself so I was hoping not to boil the ocean to get this committed. :-)
